python/graphstorm/model/gnn_encoder_base.py

```diff
@@ -210,14 +228,36 @@ def dist_inference_one_layer(layer_id, g, dataloader, target_ntypes, layer, get_
     -------
         dict of Tensors : the inferenced tensors.
     """
+    len_dataloader = max_num_batch = len(list(dataloader))
+    tensor = th.tensor([len_dataloader], device=device)
+    if is_distributed():
+        th.distributed.all_reduce(tensor, op=th.distributed.ReduceOp.MAX)
+        max_num_batch = tensor[0]
+
+    dataloader_iter = iter(dataloader)
     y = {}
-    for iter_l, (input_nodes, output_nodes, blocks) in enumerate(dataloader):
+
+    # WholeGraph does not support imbalanced batch numbers across processes/trainers
+    # TODO (IN): Fix dataloader to have same number of minibatches.
+    for iter_l in range(max_num_batch):
+        if iter_l < len_dataloader:
+            input_nodes, output_nodes, blocks = next(dataloader_iter)
+        else:
+            # Embeddings when layer_id > 0 depend on the output of layer 0, not on
+            # node features anymore. Hence, we don't need to create dummy tensors for
+            # wholegraph compatibility. Also, ntypes of input nodes might conflict with ntypes
+            # ouput nodes.
+            if int(layer_id) > 0:
+                continue
+            for ntype in g.ntypes:
+                input_nodes[ntype] = output_nodes[ntype] = th.empty((0,), dtype=g.idtype)
+            blocks = None
         if iter_l % 100000 == 0 and get_rank() == 0:
             logging.info("[Rank 0] dist_inference: finishes %d iterations.", iter_l)
 
         if task_tracker is not None:
             task_tracker.keep_alive(report_step=iter_l)
-        block = blocks[0].to(device)
+        block = blocks[0].to(device) if blocks is not None else None
```

Comment:
Let's do L261-L264 and L270 and L272 first.
Then we can skip the following steps if blocks is None.